### DISCUSSION TITLE
How can I use slash commands for discord py 2.0?

### DISCUSSION THREAD

**Discussion by Qu1ck1337** [Rapptz/discord.py#7764] [Q&A] [answered]
I can't use discord-py-slash-command, because I have some problems with import this library. So, how can I set up slash commands in by bot using discord py 2.0?

**Comment by XuaTheGrate** [comment, marked as answer]
Slash commands are available under the ``discord.app_commands`` namespace. An example command can be made as follows:
```python
import discord
from discord import app_commands

client = discord.Client(application_id=...)  # make sure to specify this ID!
tree = app_commands.CommandTree(client)

@tree.command()
@app_commands.guilds(YOUR_GUILD_ID)  # for global commands, don't use this decorator.
async def sample(interaction: discord.Interaction):
    await interaction.response.send_message("Hello!")

async def main():
    async with client:
      await tree.sync(guild=discord.Object(YOUR_GUILD_ID))  # for global commands, omit this parameter
      await client.start("YOUR_TOKEN")
```

For further reading: https://discordpy.readthedocs.io/en/master/interactions/api.html
If you have any extra questions, you are welcome to join our support server and ask us: https://discord.gg/dpy

**Reply by Qu1ck1337** [reply]
How will it works without Client just with commands.Bot() and in cogs?

**Reply by XuaTheGrate** [reply]
``commands.Bot`` works the same way, however the tree is already defined, and you can access it via ``bot.tree``.

To define app commands inside cogs, it's fairly similar to ``ext.commands``:
```python
class MyCog(commands.Cog):
    @app_commands.command()
    async def sample(self, interaction: discord.Interaction):
        pass

async def setup(bot):
    await bot.add_cog(MyCog(bot))  # this will add the commands to the tree automatically
```

**Reply by Qu1ck1337** [reply]
so I wrote a cog using `tree`
```py

import discord
from discord.ext import commands
from discord import app_commands


class Fun(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @commands.command()
    async def ping(self, ctx):
        await ctx.reply("Pong! 😄")

    @app_commands.command(name="test1")
    async def test(self, interaction: discord.Interaction):
        await interaction.response.send_message("Hello from command 1!", ephemeral=True)


async def setup(bot):
    print("Setup Fun")
    await bot.add_cog(Fun(bot))`
    
```
But I don't see slash command in slash menu and command doesn't works.
Also, main file:

```py
import asyncio
import datetime
import os
import discord
from discord.ext import commands, tasks
from config import settings, beta_settings

bot = commands.Bot(command_prefix=settings['prefix'], intents=discord.Intents.all(), help_command=None)

@bot.event
async def on_ready():
    print(f'{datetime.datetime.now().strftime("%H:%M:%S")} | [INFO] Ready!')
    update_status.start()
    
#.....

async def main():
    async with bot:
        await load_extensions()
        if settings["isBetaVersion"] is not True:
            await bot.start(settings['token'])
        else:
            await bot.start(beta_settings['beta_token'])


async def load_extensions():
    for filename in os.listdir("./cogs"):
        if filename.endswith(".py"):
            # cut off the .py from the file name
            await bot.load_extension(f"cogs.{filename[:-3]}")

asyncio.run(main())
```

**Reply by tailoric** [reply]
Two things that may be the reason:
- Global commands take up to an hour to sync. For testing define a guild command these sync immediately
- you need to call the `sync` function of your tree, discord.py doesn't auto sync (use the `guild` kwarg for syncing to your test guild) see the example from the first reply by Xua

**Reply by Qu1ck1337** [reply]
So, I get an error: discord.errors.ClientException: Client does not have an application_id set. Either the function was called before on_ready was called or application_id was not passed to the Client constructor.

I wrote bot.tree.sync() into main method:
```py
async def main():
    async with bot:
        await load_extensions()
        await bot.tree.sync(guild=discord.Object(766386682047365190))
        if settings["isBetaVersion"] is not True:
            await bot.start(settings['token'])
        else:
            await bot.start(beta_settings['beta_token'])
```

**Reply by tailoric** [reply]
please read the first reply by Xua or the error message more carefully:
```py
client = discord.Client(application_id=...)  # make sure to specify this ID!
```
you can get the id from the application page of your bot

https://discord.com/developers/applications

**Reply by Qu1ck1337** [reply]
But I have
```py
bot = commands.Bot(command_prefix=settings['prefix'], intents=discord.Intents.all(), help_command=None)
```
And I don't know how to use client and bot at the same time.

**Reply by tailoric** [reply]
bot is a client, the Bot class can do everything a client can do but also has the command extension built in.
Add the `application_id` to your Bot constructor.

```py
bot = commands.Bot(command_prefix=settings['prefix'], intents=discord.Intents.all(), help_command=None, application_id=...)
```

**Reply by Qu1ck1337** [reply]
I have an error AttributeError: '_MissingSentinel' object has no attribute 'is_set', in:
```py
    await bot.tree.sync(guild=discord.Object(766386682047365190))
```
:/

**Reply by JDJGInc** [reply]
for future questions, you should join the dpy guild or ask in there if you are in the dpy guild
